### DISCUSSION TITLE
[nix develop] user flake compilation, read-only filesystem

### DISCUSSION THREAD

**Discussion by c4lliope** [nix-community/nix-on-droid#416] [Q&A]
During a local flake compilation, some rust process ends in an error; `nix log ....` displays:

```console
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/w9lm1rcj1n9h19nrwh6fjjynqd894x8b-base64-0.1.0
source root is base64-0.1.0
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
Compiling 1 file (.ex)
[1m[32m    Updating[0m crates.io index
[1m[32m Downloading[0m crates ...
[1m[32m  Downloaded[0m lazy_static v1.4.0
[1m[31merror[0m[1m:[0m failed to download replaced source registry `crates-io`

Caused by:
  failed to create directory `/homeless-shelter/.cargo/registry/cache/index.crates.io-6f17d22bba15001f`

Caused by:
  Read-only file system (os error 30)

== Compilation error in file lib/base64.ex ==
** (RuntimeError) calling `cargo metadata` failed.

    (rustler 0.34.0) lib/rustler/compiler/config.ex:81: Rustler.Compiler.Config.metadata!/1
    (rustler 0.34.0) lib/rustler/compiler/config.ex:63: Rustler.Compiler.Config.build/1
    (rustler 0.34.0) lib/rustler/compiler.ex:8: Rustler.Compiler.compile_crate/3
    lib/base64.ex:15: (module)
```

I guess my only clue is this `/homeless-shelter` path; does this seem like a `nix-on-droid` concept? Seems like no searches are going to have any luck, especially because homeless shelters are an emergent gameplay mechanic for the Rust videogame.

I am super perplexed; besides this `nix-on-droid` seems superb and I hope I can keep up on client contracts as I travel, leaving my main computer back at home in the lab.

**Comment by t184256** [comment]
> I guess my only clue is this /homeless-shelter path; does this seem like a nix-on-droid concept?

No. There might be many meanings to 'homeless shelter', but here I'm 99% sure this specific `/homeless-shelter/` is purely a Nix concept, and it's the path that Nix sets `$HOME` to during the build process in order to ensure nothing actually uses it. If the derivation tries to, it's doing the wrong thing, so the build fails and the build instructions should be fixed.

So, something broke building of this derivation. Please ensure that the derivation in question can be built with regular Nix running on NixOS or regular Linux. https://hydra.nixos.org might be of help if you don't have access to any other PCs. Then, please report the issue to nixpkgs or wherever it is packaged. Until that's resolved, you can try to either not build it or use an older version of nixpkgs or something.

I'd try building the offending derivation somewhere else myself, but the log is too short and I have no idea what are you building. And `base64-0.1.0` is about as searcheable as homeless shelters.

> I am super perplexed; besides this nix-on-droid seems superb and I hope I can keep up on client contracts as I travel, leaving my main computer back at home in the lab.

While I'm glad you like it, this issue tracker is for nix-on-droid magic specifically and not for the whole nixpkgs made available through it.

**Comment by t184256** [comment]
... or maybe whatever you're building took a wrong turn even earlier, because [0.1.0 version of the base64 crate](https://crates.io/crates/base64/0.1.0) is 9 years old and almost certainly should not be used in 2024. Anyway, please find out what exactly in your config causes nix-on-droid to build the problematic derivation, find out where does it come from, report the issue there and either fix the issue or stay on the latest working version until it's fixed.

**Comment by c4lliope** [comment]
Yep, I'm building an elixir project that depends on a couple rust libraries, and the error always came during a rustler compilation step, building NIF definitions. `base64` had been a mix dependency.

I'm unsure the reason `base64` had been in there, none of my code really depends on this. So in the end I bypassed the `/homeless-shelter` issue, using:

```nix
    beamPackages = pkgs.beamPackages;
    deps = import ./deps.nix { inherit lib beamPackages; overrides = changes; };
    changes = (self: super: {
      egit = super.egit.override { nativeBuildInputs = with pkgs; [ libgit2.dev git ]; env.HOME = "/data/data/com.termux.nix/files/home"; };
      mdex = super.mdex.override { nativeBuildInputs = with pkgs; [ cacert cargo openssl rustc]; env.HOME = "/data/data/com.termux.nix/files/home"; };
      rustler = super.rustler.override { nativeBuildInputs = with pkgs; [ cacert cargo openssl rustc]; env.HOME = "/data/data/com.termux.nix/files/home"; };
    });
```

Specifically using the `nix-on-droid` `$HOME` seemed to erase the errors.
